drivers/serial/uart_b91.c

```diff
@@ -0,0 +1,588 @@
+{
+	uint16_t divider = 0u;
+	uint8_t bwpc = 0u;
+	const struct device *pinmux;
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	const struct uart_b91_config *cfg = GET_CFG(dev);
+
+	pinmux = DEVICE_DT_GET(DT_NODELABEL(pinmux));
+	if (!device_is_ready(pinmux)) {
+		return -ENODEV;
+	}
+
+	pinmux_pin_set(pinmux, cfg->tx_pin, cfg->tx_pin_func);
+	pinmux_pin_set(pinmux, cfg->rx_pin, cfg->rx_pin_func);
+
+	uart_b91_cal_div_and_bwpc(cfg->baud_rate, sys_clk.pclk * 1000 * 1000, &divider, &bwpc);
+	uart_b91_init(uart, divider, bwpc, UART_PARITY_NONE, UART_STOP_BIT_1);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	cfg->pirq_connect();
+#endif
+
+	return 0;
+}
+
+/* API implementation: poll_out */
+static void uart_b91_poll_out(const struct device *dev, uint8_t c)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
+	};
+
+	if (data->tx_byte_index == 0) {
+		uart->data_buf0 = c;
+	} else if (data->tx_byte_index == 1) {
+		uart->data_buf1 = c;
+	} else if (data->tx_byte_index == 2) {
+		uart->data_buf2 = c;
+	} else if (data->tx_byte_index == 3) {
+		uart->data_buf3 = c;
+	}
+
+	data->tx_byte_index++;
+	data->tx_byte_index &= 0x03;
+}
+
+/* API implementation: poll_in */
+static int uart_b91_poll_in(const struct device *dev, unsigned char *c)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	if (data->rx_byte_index == 0) {
+		*c = uart->data_buf0;
+	} else if (data->rx_byte_index == 1) {
+		*c = uart->data_buf1;
+	} else if (data->rx_byte_index == 2) {
+		*c = uart->data_buf2;
+	} else if (data->rx_byte_index == 3) {
+		*c = uart->data_buf3;
+	}
+
+	data->rx_byte_index++;
+	data->rx_byte_index &= 0x03;
+
+	return 0;
+}
+
+/* API implementation: err_check */
+static int uart_b91_err_check(const struct device *dev)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	return ((uart->status & UART_RX_ERR_STATUS) != 0) ? 1 : 0;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+/* API implementation: fifo_fill */
+static int uart_b91_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data,
+			      int size)
+{
+	int i = 0;
+
+	for (i = 0; i < size; i++) {
+		uart_b91_poll_out(dev, tx_data[i]);
+	}
+
+	return i;
+}
+
+/* API implementation: fifo_read */
+static int uart_b91_fifo_read(const struct device *dev,
+			      uint8_t *rx_data,
+			      const int size)
+{
+	int i = 0;
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	for (i = 0; i < size; i++) {
+		if (uart_b91_get_rx_bufcnt(uart) == 0) {
+			break;
+		}
+
+		uart_b91_poll_in(dev, &rx_data[i]);
+	}
+
+	return i;
+}
+
+/* API implementation: irq_tx_enable */
+static void uart_b91_irq_tx_enable(const struct device *dev)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	uart->ctrl3 = (uart->ctrl3 & (~FLD_UART_TX_IRQ_TRIQ_LEV)) |
+		      (0x01 << FLD_UART_TX_IRQ_TRIQ_LEV_OFFSET);
```

Comment:
Why not `BIT(FLD_UART_TX_IRQ_TRIQ_LEV_OFFSET)`?

Reply:
Done.